Shopify/ruby-lsp
# Limit Snippet for Cursor Positioning to VS Code

### Motivation



Resolves #1269 

The `$0` snippet, used for cursor positioning, is not universally supported across all editors. This can lead to unexpected behaviour in editors like Emacs. To address this, we can limit the use of the `$0` snippet to Visual Studio Code only.

### Implementation



1. Globally store the name of the editor via the `clientInfo.name` attribute, as defined in the [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize). The name can be surfaced in `executor.rb`, and stored in the global cache. The cache itself is defined in `store.rb`.
2. In `executor.rb`, when the routing directs to `textDocument/onTypeFormatting`, pass in the aforementioned client name.
3. Within `on_type_formatting.rb`, limit the `$0` snippet to only render when the client name is "Visual Studio Code". In essence, the cursor will only re-position itself within the block when the editor is VS Code. 

### Automated Tests



Two sets of tests have been added/updated.

In the test file for `executor.rb`, we can test that when the `clientInfo.name` attribute is provided, it is properly stored in the global cache.

In the test file for `on_type_formatting.rb`, we can test that we render the snippet appropriately based on the client name. 

### Manual Tests



1. With these changes rendered locally, you can run the LSP as is. When starting a block (e.g., `def`), you should see the `end` keyword appear automatically as expected, and the cursor should appear in the middle of the block. This suggests that this feature remains for VS Code users. See a quick demo below of this:

![cursor_vs_code](https://github.com/Shopify/ruby-lsp/assets/43973636/a08348d7-1dbd-4175-ac73-68cb7f92f876)

2. In the `move_cursor_to` method, change the first line to `return if @client_name == "Visual Studio Code"`. This line achieves the opposite of our changes, but it allows us to see the behaviour in other editors. Restart the Ruby LSP, and start a block as described previously. Now, the cursor should appear at the end of the block, since the `$0` snippet is no longer leveraged. See a quick demo below of this:

![cursor_on_non_vs_code](https://github.com/Shopify/ruby-lsp/assets/43973636/9774462c-610f-4e63-9edc-98d22c49a975)
